sql/src/main/java/org/apache/druid/sql/calcite/planner/PlannerContext.java

```diff
@@ -343,6 +349,28 @@ public String getSchemaResourceType(String schema, String resourceName)
     return plannerToolbox.rootSchema().getResourceType(schema, resourceName);
   }
 
+  /**
+   * Add a lookup name to load in the lookup loading spec.
+   */
+  public void addLookupToLoad(String lookupName)
+  {
+    if (lookupLoadingSpec.getLookupsToLoad() == null) {
+      lookupLoadingSpec = LookupLoadingSpec.loadOnly(Collections.singleton(lookupName));
+    } else {
+      Set<String> existingLookupsToLoad = new HashSet<>(lookupLoadingSpec.getLookupsToLoad());
+      existingLookupsToLoad.add(lookupName);
+      lookupLoadingSpec = LookupLoadingSpec.loadOnly(existingLookupsToLoad);
```

Comment:
So, every time we call `addLookupToLoad()`, we will end up creating a new instance of `LookupLoadingSpec`?

I suppose we don't have a choice since the set of lookups in `LookupLoadingSpec` is immutable.

Reply:
As an alternative, I suppose we could have just maintained a set of lookups to load in this class as you were doing originally.

To start with, the set would be empty and we would keep adding stuff to it in `addLookupToLoad()`.

We would still have `getLookupLoadingSpec()` where we could have this logic:

```java
      return lookupsToLoad.isEmpty() ? LookupLoadingSpec.NONE : LookupLoadingSpec.loadOnly(lookupsToLoad);
```

Reply:
@kfaraz Karan suggested that we can make `LookupLoadingSpec#lookupsToLoad` as a mutable Set, instead of the current ImmutableSet.

LookupLoadingSpec#getLookupsToLoad() can return an immutable set.

With the above changes, I'm adding the following method in LookupLoadingSpec:
```java
  public void addLookupToLoad(String lookupName)
  {
    mode = Mode.ONLY_REQUIRED;
    if (lookupsToLoad == null) {
      lookupsToLoad = new HashSet<>(Collections.singletonList(lookupName));
    } else {
      lookupsToLoad.add(lookupName);
    }
  }
```

This required making the class variables non-final, but helps in a cleaner usage everywhere else.

Hope this works.
cc: @cryptoe

Reply:
Nvm the above doesn't exactly work. We also need a way to create a new instance of PlannerContext's LookupLoadingSpec for every query.

Reply:
Have made some change, please have a look at the latest code. Thanks!